### PR TITLE
[IOTDB-5706] Data inconsistency between IoT protocol replications

### PR DESCRIPTION
**Pull request by BUAAserein**
## Description

In previous [pr](https://github.com/apache/iotdb/pull/11727), we fixed IoT Retry judgment.

In this PR, we add retry in DataRegionStateMachine write operations to fix the atomicity problem.

Please see [jira](https://issues.apache.org/jira/browse/IOTDB-5706) for details.

### CODE REVIEW

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java [hunk 1]
```diff
@@ -253,7 +253,8 @@ public static class Replication {
     private final int maxLogEntriesNumPerBatch;
     private final int maxSizePerBatch;
     private final int maxPendingBatchesNum;
-
+    private final int maxWriteRetryTime;
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java:256]: Time -> Times

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java [hunk 2]
```diff
@@ -253,7 +253,8 @@ public static class Replication {
...
+    private final long writeRetryWaitTime;
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java:257]: InMs

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java [hunk 3]
```diff
@@ -215,33 +215,54 @@ public TSStatus write(IConsensusRequest request) {
...
-        // The index is used when constructing batch in LogDispatcher. If its value
-        // increases but the corresponding request does not exist or is not put into
-        // the queue, the dispatcher will try to find the request in WAL. This behavior
-        // is not expected and will slow down the preparation speed for batch.
-        // So we need to use the lock to ensure the `offer()` and `incrementAndGet()` are
-        // in one transaction.
-        synchronized (searchIndex) {
-          logDispatcher.offer(indexedConsensusRequest);
-          searchIndex.incrementAndGet();
+
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/iot/IoTConsensusServerImpl.java:218]: I have no problem with the following logic being correct, but I have a feeling that the write function is now over 100 lines long. Maybe we can split the function into several subfunctions in this PR to make our code cleaner

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java (hunk 2 above)
Comment (BUAAserein) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java]: ok

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java (hunk 1 above)
Comment (BUAAserein) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/config/IoTConsensusConfig.java]: ok

**Review by OneSizeFitsQuorum — COMMENTED** [review]
LGTM

**Pull request merged**